### ISSUE TITLE
Class based views > function based views

### ISSUE THREAD

**Issue by valberg** [kbhff/eggplant#17]
I think we should try to avoid doing function based views and leverage the awesomeness of class based views (CBV).

I know that it can be daunting at first, but when one has played around with them a bit it because quite intuitive (in my opinion).

As an example, simple form views that require multiple lines of code in function based views take only as much as four lines with a CBV using the built-in generic CBVs:

```
class ProcessForm(FormView):
    form = SomeFormClass
    template_name = 'path/to/template.html'
    success_url = reverse('success')
```

If no one strongly disagrees with me, I will gladly convert the current function views to class based ones and write the appropriate mixins (which is also one of the cooler things about CBVs

**Comment by pawciobiel** [comment]
I think it's a matter of preference, isn't it? Let's not lock ourself to CBV idea just yet. If a function view does the job and is tested I really can't see the strong reason for it to be rewritten as CBV. 
Besides complex CBV are much more difficult for debugging.

I vote :-1:

**Comment by pawciobiel** [comment]
Although I think this is iceing on the cake I though it may make sense to use CBV in invitation especially if we go forward with the idea of account owner sending invitations to users.

https://www.pivotaltracker.com/n/projects/1337462/stories/96527974

**Comment by benjaoming** [comment]
> Besides complex CBV are much more difficult for debugging.

This is interesting. Because if you have a simple function-based view I'd say "sure whatever". But if you have something complex, I'd argue you should _always_ use a class.

Debugging is easy, I don't see how using classes makes it any harder.

**Comment by benjaoming** [comment]
I mean, the catch is: If we use functions, we get a lot of refactoring when we discover the need for reusability (and if we don't reuse, then we're just wrong).

I always do CBVs by default... no questions asked :) But I don't need a rule saying that people can't use functions and I'm not going to ask someone in a review process to refactor a view unless it's obvious that there is reusability being lost.

**Comment by pawciobiel** [comment]
I agree that finding right balance is difficult. I'm just not convinced that we have to choose either or.
My bad I don't always but it's because I can't remember CBV in v0.9(3?).

@benjaoming re debuging, let's say you have a class based view which calls `super` here and there, and somewhere, let say in a form, there is `post_save` signal called, let's throw async task somewhere too.
Are you sure it will be as easy to use PDB in CBV as it is in function views?
Race condition is another one, it is a large topic, those are always difficult to spot and debug but I think it would make it easier if it would be hidden in a function view.

**Comment by pawciobiel** [comment]
I have a strange feeling that we keep having discussions about writing code without actually writing any :dancer:
